File: examples/FSharp/OpenGL/Tutorial 1.3 - Abstractions/Shader.fs
namespace Tutorial

open System.IO
open Silk.NET.OpenGL

type Shader = {
    gl: GL
    handle: uint32
}

module Shader =
    let Created = 1
    let NotFound = -1

    // Load and compile the vertex and fragment shaders
    let private load (gl: GL) (shaderType: ShaderType) path =
        let shader = gl.CreateShader(shaderType)
        let source = File.ReadAllText(path)

        // Load and compile the shader
        gl.ShaderSource(shader, source)
        gl.CompileShader(shader)

        // Throw an exception if compiling the shader failed
        // The info log will tell us what line the shader failed on
        // so we can inspect the source.
        if gl.GetShader(shader, GLEnum.CompileStatus) <> Created then
            let info = gl.GetShaderInfoLog(shader)
            gl.DeleteShader(shader)
            failwithf "Error creating shader: %A" info

        shader


    let create (gl: GL) vertexPath fragmentPath =
        // Load and compile the shaders
        let vertexShader = load gl ShaderType.VertexShader vertexPath 
        let fragmentShader = load gl ShaderType.FragmentShader fragmentPath

        // Link the vertex and fragment shaders into a shader program
        let handle = gl.CreateProgram()
        gl.AttachShader(handle, vertexShader)
        gl.AttachShader(handle, fragmentShader)
        gl.LinkProgram(handle)

        // We're done with the shaders so they can be removed to free memory
        gl.DetachShader(handle, fragmentShader)
        gl.DetachShader(handle, vertexShader)
        gl.DeleteShader(vertexShader)
        gl.DeleteShader(fragmentShader)

        // Throw an exception if linking has failed
        if gl.GetProgram(handle, GLEnum.LinkStatus) <> Created then
            gl.GetProgramInfoLog(handle)
            |> failwithf "Error linking program %A"

        {
            gl = gl
            handle = handle
        }


    let activate (shader: Shader) =
        shader.gl.UseProgram(shader.handle)


    let setUniform (name: string) value (shader: Shader) =
        // Setting a uniform on a shader using a name
        let location = shader.gl.GetUniformLocation(shader.handle, name)
        if location = NotFound then
            failwithf "%s uniform not found on shader." name

        activate shader

        match box value with
        | :? int as i -> shader.gl.Uniform1(location, i)
        | :? float32 as f -> shader.gl.Uniform1(location, f)
        | _ -> failwithf "SetUniform cannot set value of type %A" (value.GetType())


    let destroy (shader: Shader) =
        shader.gl.DeleteProgram(shader.handle)

File: examples/CSharp/Tutorial 3.2 - Diffuse Lighting/Shader.cs
using System;
using System.IO;
using System.Numerics;
using Silk.NET.OpenGL;

namespace Tutorial
{
    public class Shader : IDisposable
    {
        private uint _handle;
        private GL _gl;

        public Shader(GL gl, string vertexPath, string fragmentPath)
        {
            _gl = gl;

            uint vertex = LoadShader(ShaderType.VertexShader, vertexPath);
            uint fragment = LoadShader(ShaderType.FragmentShader, fragmentPath);
            _handle = _gl.CreateProgram();
            _gl.AttachShader(_handle, vertex);
            _gl.AttachShader(_handle, fragment);
            _gl.LinkProgram(_handle);
            _gl.GetProgram(_handle, GLEnum.LinkStatus, out var status);
            if (status == 0)
            {
                throw new Exception($"Program failed to link with error: {_gl.GetProgramInfoLog(_handle)}");
            }

            _gl.DetachShader(_handle, vertex);
            _gl.DetachShader(_handle, fragment);
            _gl.DeleteShader(vertex);
            _gl.DeleteShader(fragment);
        }

        public void Use()
        {
            _gl.UseProgram(_handle);
        }

        public void SetUniform(string name, int value)
        {
            int location = _gl.GetUniformLocation(_handle, name);
            if (location == -1)
            {
                throw new Exception($"{name} uniform not found on shader.");
            }
            _gl.Uniform1(location, value);
        }

        public unsafe void SetUniform(string name, Matrix4x4 value)
        {
            //A new overload has been created for setting a uniform so we can use the transform in our shader.
            int location = _gl.GetUniformLocation(_handle, name);
            if (location == -1)
            {
                throw new Exception($"{name} uniform not found on shader.");
            }
            _gl.UniformMatrix4(location, 1, false, (float*) &value);
        }

        public void SetUniform(string name, float value)
        {
            int location = _gl.GetUniformLocation(_handle, name);
            if (location == -1)
            {
                throw new Exception($"{name} uniform not found on shader.");
            }
            _gl.Uniform1(location, value);
        }

        public void SetUniform(string name, Vector3 value)
        {
            int location = _gl.GetUniformLocation(_handle, name);
            if (location == -1)
            {
                throw new Exception($"{name} uniform not found on shader.");
            }
            _gl.Uniform3(location, value.X, value.Y, value.Z);
        }

        public void Dispose()
        {
            _gl.DeleteProgram(_handle);
        }

        private uint LoadShader(ShaderType type, string path)
        {
            string src = File.ReadAllText(path);
            uint handle = _gl.CreateShader(type);
            _gl.ShaderSource(handle, src);
            _gl.CompileShader(handle);
            string infoLog = _gl.GetShaderInfoLog(handle);
            if (!string.IsNullOrWhiteSpace(infoLog))
            {
                throw new Exception($"Error compiling shader of type {type}, failed with error {infoLog}");
            }

            return handle;
        }
    }
}
